### PR TITLE
TPM2_DictionaryAttackLockReset & TPM2_DictionaryAttackParameters support

### PR DESCRIPTION
**Pull request by t33m**
added support for TPM2_DictionaryAttackLockReset and TPM2_DictionaryAttackParameters commands

### CODE REVIEW

**Comment by t33m** [comment]
@googlebot I signed it

File: tpm2/test/tpm2_test.go [hunk 1]
```diff
@@ -1656,3 +1656,117 @@ func TestPolicyPCR(t *testing.T) {
...
+	if _, err = RSADecrypt(rw, handle, "bad password", encrypted, scheme, label); err == nil {
+		t.Fatal("RSADecrypt with bad password must fail")
+	}
+
+	caps, _, err := GetCapability(rw, CapabilityTPMProperties, 1, uint32(LockoutCounter))
+	if err != nil {
+		t.Fatalf("GetCapability failed: %v", err)
+	}
+	if caps[0].(TaggedProperty).Value != 1 {
+		t.Fatalf("expected 1, but got %d", caps[0].(TaggedProperty).Value)
```
Comment (twitchy-jsonp) [inline tpm2/test/tpm2_test.go]: Please re-order the message so the actual value is first, and the expected value is second (ie: [got-before-want](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures) idiom)

File: tpm2/test/tpm2_test.go [hunk 2]
```diff
@@ -1656,3 +1656,117 @@ func TestPolicyPCR(t *testing.T) {
...
+	if err = DictionaryAttackLockReset(rw, auth); err != nil {
+		t.Fatalf("DictionaryAttackLockReset failed: %v", err)
+	}
+
+	caps, _, err = GetCapability(rw, CapabilityTPMProperties, 1, uint32(LockoutCounter))
+	if err != nil {
+		t.Fatalf("GetCapability failed: %v", err)
+	}
+	if caps[0].(TaggedProperty).Value != 0 {
+		t.Fatalf("expected 0, but got %d", caps[0].(TaggedProperty).Value)
```
Comment (twitchy-jsonp) [inline tpm2/test/tpm2_test.go]: Ditto [got-before-want](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures) plz

File: tpm2/tpm2.go [hunk 3]
```diff
@@ -1838,3 +1838,44 @@ func RSADecrypt(rw io.ReadWriter, key tpmutil.Handle, password string, message [
...
+	}
+	cmd, err := concat(ha, encodedAuth)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, TagSessions, cmdDictionaryAttackLockReset, tpmutil.RawBytes(cmd))
+	return err
+}
+
+// DictionaryAttackParameters changes the lockout parameters
```
Comment (twitchy-jsonp) [inline tpm2/tpm2.go]: Make this a full sentence, and maybe describe what kind of authorization is typically needed?

File: tpm2/test/tpm2_test.go [hunk 4]
```diff
@@ -1656,3 +1656,117 @@ func TestPolicyPCR(t *testing.T) {
...
+		t.Run(tc.name, func(t *testing.T) {
+			if err := DictionaryAttackParameters(rw, auth, tc.maxTries, tc.recoveryTime, tc.lockoutRecovery); err != nil {
+				t.Fatalf("DictionaryAttackParameters failed: %v", err)
+			}
+			caps, _, err := GetCapability(rw, CapabilityTPMProperties, 3, uint32(MaxAuthFail))
+			if err != nil {
+				t.Fatalf("GetCapability failed: %v", err)
+			}
+			if tc.maxTries != caps[0].(TaggedProperty).Value {
+				t.Fatalf("expected %d, but got %d", tc.maxTries, caps[0].(TaggedProperty).Value)
```
Comment (twitchy-jsonp) [inline tpm2/test/tpm2_test.go]: Ditto [got-before-want](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures) failure message order plz, here and below

File: tpm2/constants.go [hunk 5]
```diff
@@ -236,6 +236,31 @@ const (
...
+	AlgorithmSet
+	LoadedCurves
+	LockoutCounter
+	MaxAuthFail
+	LockoutInterval
+	LockoutRecovery
+	NVWriteRecovery
+	AuditCounter0
+	AuditCounter1
+)
```
Comment (josephlr) [inline tpm2/constants.go]: I agree that we want to add these, but we should format this section in a different way (both for Godoc and for people reading the code).

First, change the statement for `TPMProp` to make it clear what this property actually is:
```go
// TPMProp represents a Property Tag (TPM_PT) used with calls to GetCapability(CapabilityTPMProperties).
type TPMProp uint32
``` 
Next, we have a section describing the Fixed TPM Properties:
```go
// TPM Capability Properties, see TPM 2.0 Spec, Rev 1.38, Table 23.
// Fixed TPM Properties (PT_FIXED)
const (
	FamilyIndicator TPMProp = 0x100 + iota
	...
	CapabilityMaxBufferSize
)
```
Then, we have a section describing the Variable TPM Properties:
```go
// Variable TPM Properties (PT_VAR)
const (
	TPMAPermanent     TPMProp = 0x200 + iota
	...
	AuditCounter1
)
```
Finally, a section for things we gave type `TPMProp` incorrectly:
```go
// Allowed ranges of different kinds of Handles (TPM_HANDLE)
// These constants have type TPMProp for backwards compatibility.
const (
	PCRFirst           TPMProp = 0x00000000
	...
	PermanentLast      TPMProp = 0x4000010F
)

```

File: tpm2/constants.go [hunk 6]
```diff
@@ -236,6 +236,31 @@ const (
 	PermanentLast      TPMProp = 0x4000010F
 )
 
+const (
+	// Fixed TPM Properties (PT_VAR)
```
Comment (josephlr) [inline tpm2/constants.go]: Variable TPM Properties

File: tpm2/tpm2.go [hunk 7]
```diff
@@ -1838,3 +1838,44 @@ func RSADecrypt(rw io.ReadWriter, key tpmutil.Handle, password string, message [
 	}
 	return decodeRSADecrypt(resp)
 }
+
+// DictionaryAttackLockReset cancels the effect of a TPM lockout due to a number of successive authorization failures.
+// If this command is properly authorized, the lockout counter is set to zero.
```
Comment (josephlr) [inline tpm2/tpm2.go]: We can just have the description here be:
```go
// DictionaryAttackLockReset cancels the effect of a TPM lockout due to a number
// of successive authorization failures, by setting the lockout counter to zero.
```

File: tpm2/tpm2.go (hunk 3 above)
Comment (josephlr) [reply tpm2/tpm2.go]: If we describe the auth concerns in `DictionaryAttackLockReset`, we can just say that the auth for this command is the same, and leave it at that.

File: tpm2/tpm2.go [hunk 8]
```diff
@@ -1838,3 +1838,44 @@ func RSADecrypt(rw io.ReadWriter, key tpmutil.Handle, password string, message [
 	}
 	return decodeRSADecrypt(resp)
 }
+
```
Comment (josephlr) [inline tpm2/tpm2.go]: We should also talk about authorization for this command, specifically that:
  - We need to use the "Lockout Authorization" value.
  - The value defaults to the empty password session auth.
  - That this value is set via a call to `HierarchyChangeAuth(HandleLockout)`

**Comment by t33m** [comment]
@josephlr, @twitchy-jsonp, plz check now

**Comment by t33m** [comment]
@twitchy-jsonp , @josephlr , hi, folks! still waiting your review!

**Comment by t33m** [comment]
@twitchy-jsonp , @josephlr , could you merge this?

**Pull request merged**